base/reduce.jl

```diff
@@ -452,28 +461,29 @@ _mapreduce(f, op, ::IndexCartesian, A::AbstractArrayOrBroadcasted) = mapfoldl(f,
 """
     reduce(op, itr; [init])
 
-Reduce the given collection `itr` with the given binary operator `op`. If provided, the
-initial value `init` must be a neutral element for `op` that will be returned for empty
-collections. It is unspecified whether `init` is used for non-empty collections.
+Reduce the given collection `itr` with the given binary operator `op`.
+
+The order of evaluations and the associativity of the reduction is unspecified.
+This means that you shouldn't
+use non-associative operations like `-` because it is undefined whether `reduce(-,[1,2,3])`
+will be evaluated as `(1-2)-3` or `1-(2-3)`. Use
+[`foldl`](@ref) or [`foldr`](@ref) for guaranteed left or right associativity.
 
-For empty collections, providing `init` will be necessary, except for some special cases
-(e.g. when `op` is one of `+`, `*`, `max`, `min`, `&`, `|`) when Julia can determine the
-neutral element of `op`.
+If provided, `init` serves as the return value for empty `itrs`. For non-empty iterators,
+it is included in the reduction exactly once and ensures that every element in `itrs` is
+used as an argument to `op`. Like the reduction itself, the exact order and associativity
+of how `init` is included is not specified. It is generally an error to call `mapreduce`
+with empty collections without specifying an `init` value, but in unambiguous cases the
+identity value may be returned; see [`Base.reduce_empty`](@ref) for more details.
 
 Reductions for certain commonly-used operators may have special implementations, and
 should be used instead: [`maximum`](@ref)`(itr)`, [`minimum`](@ref)`(itr)`, [`sum`](@ref)`(itr)`,
 [`prod`](@ref)`(itr)`, [`any`](@ref)`(itr)`, [`all`](@ref)`(itr)`.
 There are efficient methods for concatenating certain arrays of arrays
 by calling `reduce(`[`vcat`](@ref)`, arr)` or `reduce(`[`hcat`](@ref)`, arr)`.
 
-The associativity of the reduction is implementation dependent. This means that you can't
-use non-associative operations like `-` because it is undefined whether `reduce(-,[1,2,3])`
-should be evaluated as `(1-2)-3` or `1-(2-3)`. Use [`foldl`](@ref) or
-[`foldr`](@ref) instead for guaranteed left or right associativity.
-
 Some operations accumulate error. Parallelism will be easier if the reduction can be
-executed in groups. Future versions of Julia might change the algorithm. Note that the
-elements are not reordered if you use an ordered collection.
```

Comment:
I have no idea what this means?

Reply:
I think the point it is trying to make is that:
1. the associativity of `mapreduce` is not specified — e.g. `mapreduce(+, [a,b,c,d])` might be computed as `a + (b + (c + d))` or `((a + b) + c) + d` or `(a + b) + (c + d)`.     If these are floating-point operations, different associativities can change the roundoff errors.   Future versions of Julia might change the associativity for reasons of performance, accuracy (e.g. [pairwise reductions](https://en.wikipedia.org/wiki/Pairwise_summation) are already used in some cases to improve accuracy), or parallelism (though `mapreduce` is currently serial).
2. for an ordered collection (e.g. an array), although the associativity can change the ordering cannot — Julia does not assume commutativity.   For an unordered collection (e.g. a `Set`), the ordering is also implementation defined (so the results could completely change from version to version if `op` is not commutative).

Reply:
Ohhh I see. I'm so used to associative operators being commutative I didn't even consider that possibility.  But of course.  Any suggestions on how we might make this clearer?  It's something that I could easily see a custom implementation getting wrong — although fortunately the "simple" algorithms one might write probably "happen" to do this correctly.

I had actually used the word "order" above to specify that the sequence of when `f` gets applied to each element in `itr` is not defined.

Reply:
Maybe talk about the "operand ordering".